File: src/RecursiveFactorial.php
<?php

namespace Victor\GrokkingAlgorithmsExercises;

final class RecursiveFactorial
{
    public function execute(int $number): int
    {
        if ($number == 1) {
            return 1;
        } else {
            return $number * $this->execute($number - 1);
        }
    }
}

File: lib/grokking_algorithms_exercises/recursive_factorial.rb
# frozen_string_literal: true

module GrokkingAlgorithmsExercises
  # Factorial of a number through recursion, Big O execution time: O(n)
  class RecursiveFactorial
    def execute(number)
      raise ArgumentError, "number must not be negative" if number.negative?

      return 1 if number <= 1

      number * execute(number - 1)
    end
  end
end
